bsp/nrf/lh2.c

```diff
@@ -263,102 +802,130 @@ void db_lh2_init(db_lh2_t *lh2, const gpio_t *gpio_d, const gpio_t *gpio_e) {
-        lh2->raw_data[location].bit_offset           = 0;
-        lh2->locations[location].selected_polynomial = LH2_POLYNOMIAL_ERROR_INDICATOR;
-        lh2->locations[location].lfsr_location       = LH2_LOCATION_ERROR_INDICATOR;
-        //_lh2_vars.data[location].envelope_duration   = 0xFFFFFFFF;
-        memset(_lh2_vars.data[location].buffer, 0, LH2_BUFFER_SIZE);
+    lh2->spi_ring_buffer_count_ptr = &_lh2_vars.data.count;  // pointer to the size of the spi ring buffer,
+
+    for (uint8_t sweep = 0; sweep < LH2_SWEEP_COUNT; sweep++) {
+        for (uint8_t basestation = 0; basestation < LH2_SWEEP_COUNT; basestation++) {
+            lh2->raw_data[sweep][basestation].bits_sweep           = 0;
+            lh2->raw_data[sweep][basestation].selected_polynomial  = LH2_POLYNOMIAL_ERROR_INDICATOR;
+            lh2->raw_data[sweep][basestation].bit_offset           = 0;
+            lh2->locations[sweep][basestation].selected_polynomial = LH2_POLYNOMIAL_ERROR_INDICATOR;
+            lh2->locations[sweep][basestation].lfsr_location       = LH2_LOCATION_ERROR_INDICATOR;
+            lh2->timestamps[sweep][basestation]                    = 0;
+            lh2->data_ready[sweep][basestation]                    = DB_LH2_NO_NEW_DATA;
+        }
     }
+    memset(_lh2_vars.data.buffer[0], 0, LH2_BUFFER_SIZE);
+
+    // Initialize the hash table for the lsfr checkpoints
+    _fill_hash_table(_end_buffers_hashtable);
 
     // initialize GPIOTEs
     _gpiote_setup(gpio_e);
 
     // initialize PPI
     _ppi_setup();
-
-    lh2->state = DB_LH2_IDLE;
 }
 
-void db_lh2_start(db_lh2_t *lh2) {
-    db_lh2_reset(lh2);
-    NRF_PPI->CHENSET = (1 << PPI_SPI_START_CHAN) | (1 << PPI_SPI_STOP_CHAN);
+void db_lh2_start(void) {
 
-    lh2->state = DB_LH2_RUNNING;
+    NRF_PPI->TASKS_CHG[PPI_SPI_GROUP].EN = 1;
 }
 
-void db_lh2_stop(db_lh2_t *lh2) {
-    NRF_PPI->CHENCLR = (1 << PPI_SPI_START_CHAN) | (1 << PPI_SPI_STOP_CHAN);
-    lh2->state       = DB_LH2_IDLE;
+void db_lh2_stop(void) {
+
+    NRF_PPI->TASKS_CHG[PPI_SPI_GROUP].DIS = 1;
 }
 
 void db_lh2_reset(db_lh2_t *lh2) {
-    _lh2_vars.transfer_counter   = 0;
-    _lh2_vars.lha_packet_counter = 0;
-    _lh2_vars.lhb_packet_counter = 0;
-    _lh2_vars.buffers_ready      = false;
-    lh2->state                   = DB_LH2_RUNNING;
+    // compute LFSR locations and detect invalid packets
+    for (uint8_t basestation = 0; basestation < LH2_BASESTATION_COUNT; basestation++) {
+        for (uint8_t sweep = 0; sweep < 2; sweep++) {
+
+            // Remove the flags indicating available data
+            lh2->data_ready[sweep][basestation] = DB_LH2_NO_NEW_DATA;
+            lh2->timestamps[sweep][basestation] = 0;
+            // We won't actually clear the data, it's not worth the computational effort.
+        }
+    }
 }
 
 void db_lh2_process_raw_data(db_lh2_t *lh2) {
-    if (!_lh2_vars.buffers_ready) {
+    if (_lh2_vars.data.count <= 0) {
```

Comment:
count is `uint8_t`, just do:
```suggestion
    if (_lh2_vars.data.count == 0) {
```

Reply:
done